feat: Add prometheus deployment manifests to setup tools

bradbeam opened the pull request:
Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
feat: Add prometheus deployment manifests to setup tools

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on config/docker-entrypoint.sh:
```diff
@@ -1,19 +1,22 @@
+  ;;
+  prometheus)
+    # Generate prometheus manifests
+    shift && cd /workspace/rso
```
This should be `/workspace/prometheus` you can just wipe out the existing service monitor (the endpoints it would have been using are gone, it's dead code).

jgustie commented on config/docker-entrypoint.sh:
```diff
@@ -31,6 +34,8 @@ fi
 
 # Add trial labels to the resulting manifests so they can be more easily located
 if [ -n "$TRIAL" ]; then
+    # Note, this heredoc block must be indented with tabs
+    # <<- allows for indentation via tabs, if spaces are used it is no good.
```
I'm sensing you got burned by this?

bradbeam replied:
yup. saw the indentation looked off, #thereifixedit and it no longer worked.

jgustie commented on config/docker-entrypoint.sh:
```diff
@@ -44,12 +49,29 @@ if [ -n "$TRIAL" ]; then
+		  name: experiment-labels
+		labels:
+		  "redskyops.dev/experiment": $EXPERIMENT
+		  "redskyops.dev/experiment": experimentResource
```
This second (duplicate) label should be `experiment-role`; also you should add it to `well_known_labels.go`.

jgustie commented on config/docker-entrypoint.sh:
```diff
@@ -44,12 +49,29 @@ if [ -n "$TRIAL" ]; then
     create)
         handle () {
-            kubectl create -f -
+            kubectl apply -f -
```
This was explicitly `create` for `generate-name` and to force failure on conflict. You should add another case for `apply` (it used to be there but got removed after we weren't using it for anything).

jgustie commented on config/rso/kube-state-metrics-deployment.yaml:
```diff
@@ -0,0 +1,52 @@
+kind: Deployment
+metadata:
+  name: kube-state-metrics
+  namespace: default
```
We shouldn't be putting things in the default namespace like this: I would have expected the namespace to be omitted from all of the resources (even if the edit at runtime is going to add a namespace transformer).

Also, even though the entry point script should create the right kustomization by scanning for resources, it should still be explicitly included in the directory and should still explicitly list all of the resources).

jgustie commented on config/rso/prometheus-server-rbac.yaml:
```diff
@@ -0,0 +1,63 @@
+#      - "/metrics"
+#    verbs:
+#      - get
+#
```
Remove these comments, do it in a separate commit if you might want them back.

bradbeam committed:
refactor: Move prometheus deployment from rso to prometheus directory

Since we arent using the kubebuilder servicemonitor bits anymore, we can repurpose this directory.

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
refactor: Remove default namespace from resources

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Add explicit emptyDir for prometheus deployment

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

bradbeam committed:
fix: Handle existing kustomization file

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>

jgustie commented on config/prometheus/prometheus-server-service.yaml:
```diff
@@ -0,0 +1,35 @@
+metadata:
+  labels:
+    app: prometheus
+  name: prometheus-server
```
Can we just name this `prometheus` so when we have URL support we can have a default of `http://prometheus:9090` ?

bradbeam replied:
Want to avoid `prometheus` since that has a potential for conflict. Will use `rso-prometheus` below and remove this one.

bradbeam replied:
along that vein I should probably change all the basic 'prometheus' names to something more specific. ugh.

jgustie replied:
I like the change of the name prefix for all the other objects, I still don't like it on the service because of the whole sensible defaults thing.

A default of "http://prometheus:9090" is a good default because it is likely to actually hit (our Prometheus or theirs, it's generic); if the service name is "redsky-prometheus" then we either need to make the default URL "http://redsky-prometheus:9090" (only likely to hit our Prometheus) or stick with "http://prometheus:9090" and have a default that needs to be overwritten in the particular use case we established the default behavior for in the first place.

I'm fine to revisit this once we are closer with URL based metric collection.

jgustie commented on config/prometheus/prometheus-server-service.yaml:
```diff
@@ -0,0 +1,35 @@
+  selector:
+    app: prometheus
+  sessionAffinity: None
+  type: ClusterIP
```
Is this a duplicate service?

jgustie commented on config/prometheus/prometheus-server-deployment.yaml:
```diff
@@ -0,0 +1,76 @@
+              readOnly: true
+        - name: prometheus-server
+          image: "prom/prometheus:latest"
+          imagePullPolicy: "IfNotPresent"
```
I would be a little concerned trusting `latest` here, can we pin this to a version (at least a major, maybe major/minor)?

bradbeam committed:
refactor: Add `redsky-` prefix to prometheus object

Signed-off-by: Brad Beam <brad.beam@carbonrelay.com>
